// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: envoy/extensions/upstreams/http/generic/v3/generic_connection_pool.proto

package genericv3

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on GenericConnectionPoolProto with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GenericConnectionPoolProto) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GenericConnectionPoolProto with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GenericConnectionPoolProtoMultiError, or nil if none found.
func (m *GenericConnectionPoolProto) ValidateAll() error {
	return m.validate(true)
}

func (m *GenericConnectionPoolProto) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return GenericConnectionPoolProtoMultiError(errors)
	}

	return nil
}

// GenericConnectionPoolProtoMultiError is an error wrapping multiple
// validation errors returned by GenericConnectionPoolProto.ValidateAll() if
// the designated constraints aren't met.
type GenericConnectionPoolProtoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GenericConnectionPoolProtoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GenericConnectionPoolProtoMultiError) AllErrors() []error { return m }

// GenericConnectionPoolProtoValidationError is the validation error returned
// by GenericConnectionPoolProto.Validate if the designated constraints aren't met.
type GenericConnectionPoolProtoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GenericConnectionPoolProtoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GenericConnectionPoolProtoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GenericConnectionPoolProtoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GenericConnectionPoolProtoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GenericConnectionPoolProtoValidationError) ErrorName() string {
	return "GenericConnectionPoolProtoValidationError"
}

// Error satisfies the builtin error interface
func (e GenericConnectionPoolProtoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGenericConnectionPoolProto.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GenericConnectionPoolProtoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GenericConnectionPoolProtoValidationError{}
